File: C++/call_by/call_by.cpp
#include <iostream>
#include <string>
#include <vector>
using std::vector;
using std::cout;
using std::endl;
using std::string;
void print(string name, vector<int> list) {
  cout<<name<<endl;
  for(int ii = 0;ii < list.size();ii++) {
    cout<<list.at(ii)<<endl;
  }
}
void call_by_value(vector<int> list){
  list.push_back(1);
}
void call_by_value2(vector<int> list){
  list = vector<int>();
  list.push_back(1);
}
void call_by_reference(vector<int>& list){
  list.push_back(1);
}
void call_by_reference2(vector<int>& list){
  list = vector<int>();
  list.push_back(1);
}
void call_by_sharing(vector<int>* list){
  list->push_back(1);
}
void call_by_sharing2(vector<int>* list){
  list = new vector<int>();
  list->push_back(1);
}

int main(void) {
  vector<int> list;

  list = vector<int>();
  list.push_back(0);
  call_by_value(list);
  print("call by value", list);

  list = vector<int>();
  list.push_back(0);
  call_by_value2(list);
  print("call by value2", list);

  list = vector<int>();
  list.push_back(0);
  call_by_reference(list);
  print("call by reference", list);

  list = vector<int>();
  list.push_back(0);
  call_by_reference2(list);
  print("call by reference2", list);

  list = vector<int>();
  list.push_back(0);
  call_by_sharing(&list);
  print("call by sharing", list);

  list = vector<int>();
  list.push_back(0);
  call_by_sharing2(&list);
  print("call by sharing2", list);
}

File: c#/CallBy/CallBy.cs
using System.Collections.Generic;
class CallBy
{
    static void Main(string[] args)
    {
        List<int> list ;

        list = new List<int>();
        list.Add(0);
        CallByReference(ref list);
        Print("Call by Reference", list);

        list = new List<int>();
        list.Add(0);
        CallByReference2(ref list);
        Print("Call by Reference2", list);

        list = new List<int>();
        list.Add(0);
        CallBySharing(list);
        Print("Call by Sharing", list);

        list = new List<int>();
        list.Add(0);
        CallBySharing2(list);
        Print("Call by Sharing2", list);
    }
    static void CallByReference(ref List<int> list)
    {
        list.Add(1);
    }
    static void CallByReference2(ref List<int> list)
    {
        list = new List<int>();
        list.Add(1);
    }
    static void CallBySharing(List<int> list)
    {
        list.Add(1);
    }
    static void CallBySharing2(List<int> list)
    {
        list = new List<int>();
        list.Add(1);
    }
    static void Print(string name, List<int> list)
    {
        System.Console.WriteLine(name);
        foreach (int element in list)
        {
            System.Console.WriteLine(element);
        }
    }
}